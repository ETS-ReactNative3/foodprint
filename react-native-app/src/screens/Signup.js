import React, { useState } from 'react';
import { View } from 'react-native';
import { Button, Input, Text } from 'react-native-elements';
import { ScrollView } from 'react-native-gesture-handler';

export default Signup = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  const EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

  const signUpHandler = () => {
    if (!EMAIL_REGEX.test(email)) {
      return console.warn('alert user, email invalid.');
    }

    if (password.length < 8) {
      return console.warn('alert user, password too short ');
    }

    // TODO add more password security checks
    // TODO change view to move button up when keyboard is shown

    const userDetails = {
      email,
      password,
    };
    console.log({ userDetails });
  };

  return (
    <ScrollView contentContainerStyle={{ alignItems: 'center' }} style={{ width: '100%', height: '100%' }}>
      <View style={{ alignItems: 'center', justifyContent: 'center', marginVertical: 50 }}>
        <Text h1>
          FoodPrint
          </Text>
      </View>
      <View style={{ width: '80%' }}>
        <View>
          <Input label={'Your Email Address'} placeholder="banana@foodprint.co" value={email} onChangeText={value => setEmail(value)} />
          <Input label={'Password'} placeholder="Password" value={password} onChangeText={value => setPassword(value)} secureTextEntry={true} />
        </View>
      </View>
      <View style={{ width: '80%' }}>
        <Button
          buttonStyle={{ backgroundColor: 'green', marginVertical: 100 }}
          titleStyle={{ fontSize: 24 }}
          title="Register"
          onPress={signUpHandler}
        />
      </View>
    </ScrollView>
  );
};
