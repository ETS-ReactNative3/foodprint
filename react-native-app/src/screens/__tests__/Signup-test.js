import React from 'react';
import { fireEvent, render, wait } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import Signup from '../Signup';
import { AuthContext } from '../../store/Auth';

// Snapshot test
test('SignUp renders matches previous snapshot', () => {
  const authentication = {
    signUp: () => { },
  };

  const tree = renderer.create(<AuthContext.Provider value={authentication}>
    <Signup />
  </AuthContext.Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Join button calls signUp function from context if email and password is valid', async () => {
  const signUpMock = jest.fn();

  const authentication = {
    signUp: signUpMock,
  };

  const utils = renderSignUp(authentication);

  const emailInput = utils.getByPlaceholderText('Email');
  const famousLecturerAtImperial = 'WillKnottenbelt@NotARealEmail.co.uk';
  fireEvent.changeText(emailInput, famousLecturerAtImperial);

  const passwordInput = utils.getByPlaceholderText('Password');
  const famousLecturersPassword = 'b10ckch@1n';
  fireEvent.changeText(passwordInput, famousLecturersPassword);

  const button = utils.getByText('Join');
  fireEvent.press(button);
  await wait(() => expect(utils.queryByTestId('joinButton')));
  expect(signUpMock.mock.calls.length).toBe(1);
});

// TODO implement error messages for user
// test('Prevents badly formatted email', async () => {
//   const signUpMock = jest.fn();

//   const authentication = {
//     signUp: signUpMock,
//   };

//   const utils = renderSignUp(authentication);

//   const emailInput = utils.getByPlaceholderText('Email');
//   const invalidEmail = 'NotAnEmail';
//   fireEvent.changeText(emailInput, invalidEmail);

//   const passwordInput = utils.getByPlaceholderText('Password');
//   const password = 'b10ckch@1n';
//   fireEvent.changeText(passwordInput, password);

//   const button = utils.getByText('Join');
//   fireEvent.press(button);
//   await wait(() => expect(utils.queryByTestId('joinButton')));
// });


// HELPERS

// Re-use this function to render Signup, as it uses useContext. This 'mocks' that out.
function renderSignUp(authentication) {
  return render(
    <AuthContext.Provider value={authentication}>
      <Signup />
    </AuthContext.Provider>
  );
}

