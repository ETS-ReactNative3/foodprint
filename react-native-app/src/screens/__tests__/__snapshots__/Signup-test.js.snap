// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp renders matches previous snapshot 1`] = `
<SafeAreaView>
  <ScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
      }
    }
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginVertical": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 50,
          }
        }
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
      >
        FoodPrint
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "80%",
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "marginBottom": 15,
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autoCorrect={false}
              editable={true}
              onChangeText={[Function]}
              onSubmitEditing={[Function]}
              placeholder="Email"
              rejectResponderTermination={true}
              returnKeyType="next"
              style={
                Object {
                  "alignSelf": "center",
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "minHeight": 40,
                }
              }
              testID="RNE__Input__text-input"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 15,
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autoCorrect={false}
              editable={true}
              onChangeText={[Function]}
              onSubmitEditing={[Function]}
              placeholder="Password"
              rejectResponderTermination={true}
              returnKeyType="join"
              secureTextEntry={true}
              style={
                Object {
                  "alignSelf": "center",
                  "color": "black",
                  "flex": 1,
                  "fontSize": 18,
                  "minHeight": 40,
                }
              }
              testID="RNE__Input__text-input"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          style={
            Object {
              "borderRadius": 3,
            }
          }
        >
          <TouchableOpacity
            activeOpacity={0.3}
            disabled={false}
            onPress={[Function]}
            testID="joinButton"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "green",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginVertical": 16,
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontSize": 24,
                    "paddingBottom": 1,
                    "paddingTop": 2,
                    "textAlign": "center",
                  }
                }
              >
                Join
              </Text>
            </View>
          </TouchableOpacity>
        </View>
      </View>
    </View>
  </ScrollView>
</SafeAreaView>
`;
